jbpm-system-test/jbpm-smartrouter-test/src/test/java/org/kie/samples/integration/SmartRouterSystemTest.java

```diff
@@ -0,0 +1,445 @@
+                         Arguments.of(null, limitData3),
+                         Arguments.of(null, resetPeer3),
+                         Arguments.of(null, timeout3),
+                         Arguments.of(limitData1, limitData3),
+                         Arguments.of(limitData1, resetPeer3),
+                         Arguments.of(limitData1, timeout3),
+                         Arguments.of(resetPeer1, limitData3),
+                         Arguments.of(resetPeer1, resetPeer3),
+                         Arguments.of(resetPeer1, timeout3),
+                         Arguments.of(timeout1, limitData3),
+                         Arguments.of(timeout1, resetPeer3),
+                         Arguments.of(timeout1, timeout3),
+                         Arguments.of(null, null));
+    }
+    
+    public static int getRandomTimeout(int min, int max) {
+        Random random = new Random();
+        int timeout = random.ints(min, max).findFirst().getAsInt();
+        logger.debug("random timeout is {} ms", timeout);
+        return timeout;
+    }
+    
+    @ParameterizedTest
+    @MethodSource("provideToxics")
+    void shouldUpdateRoutingTableWhenDisconnectedAndReconnected(ToxicSupplier<Toxic, IOException> toxic1, 
+                                                                ToxicSupplier<Toxic, IOException> toxic3) throws Exception {
+        logger.debug("*** Cut Connection ***");
+        
+        cutConnection(toxic1, toxic3);
+        
+        logger.debug("*** Reestablish Connection ***");
+        removeAllToxics(proxy1);
+        removeAllToxics(proxy3);
+        
+        await().atMost(ONE_MINUTE).untilAsserted(() -> assertThat(readNumberOfContainers()).isEqualTo(2));
+        
+        await().atMost(ONE_MINUTE).until(() -> assertServersInRoutingTable(URL_NODE1, URL_NODE3));
+    }
+
+    @ParameterizedTest
+    @MethodSource("provideToxics")
+    void shouldStopPollingWhenReconnectionLimitIsReached(ToxicSupplier<Toxic, IOException> toxic1, 
+                                                         ToxicSupplier<Toxic, IOException> toxic3) throws Exception {
+        
+        logger.debug("*** Cut Connection ***");
+        TestFixture fixture = cutConnection(toxic1, toxic3);
+        
+        if (!fixture.exercisedToxics.isEmpty()) {
+            logger.debug("*** Waiting for max_reach ***");
+            consumerSmartRouter.waitUntil(frame -> frame.getUtf8String().contains(REACHED_ATTEMPTS_LIMIT), 
+                50, TimeUnit.SECONDS, 2-fixture.expectedContainers+reachedLimitCounter);
+        }
+        
+        logger.debug("*** Reestablish Connection ***");
+        removeAllToxics(proxy1);
+        removeAllToxics(proxy3);
+        
+        await().atMost(ONE_MINUTE).untilAsserted(() -> assertThat(readNumberOfContainers()).isEqualTo(fixture.expectedContainers));
+        
+        await().atMost(ONE_MINUTE).until(() -> assertServersInRoutingTable(fixture.expectedServers[1], fixture.expectedServers[3]));
```

Comment:
if we are removing all toxics, shouldn't we have all servers back up and available. i.e in our case, all 2 servers should be available (number of expected containers to be 2). Or am I missing anything else?

Reply:
expectedServers is an array, and positions 1 and 3 are fulfilled (corresponding respectively to nodes 1 and 3)